[docs] refactor docs for easier info parsing

sayakpaul opened the pull request:
Approach taken:

* Keep the README as less cluttered as possible. 
* Add docs for each model. Only when we have more training algos can we include separate sections within model docs.
* Once we have I2V support for a model that should go into the respective model accordingly.
* Have a central doc for memory optimizations.

sayakpaul commented on README.md:
````diff
@@ -128,540 +125,24 @@ eval $cmd
 echo -ne "-------------------- Finished executing script --------------------\n\n"
 ```
 
-### Inference:
````
Went to specific model docs.

sayakpaul commented on README.md:
```diff
@@ -128,540 +125,24 @@ eval $cmd
 
-If you would like to use a custom dataset, refer to the dataset preparation guide [here](./assets/dataset.md).
-
-> [!NOTE]
-> To lower memory requirements:
-> - Use a DeepSpeed config to launch training (refer to [`accelerate_configs/deepspeed.yaml`](./accelerate_configs/deepspeed.yaml) as an example).
-> - Pass `--precompute_conditions` when launching training.
-> - Pass `--gradient_checkpointing` when launching training.
-> - Pass `--use_8bit_bnb` when launching training. Note that this is only applicable to Adam and AdamW optimizers.
-> - Do not perform validation/testing. This saves a significant amount of memory, which can be used to focus solely on training if you're on smaller VRAM GPUs.
```
In `docs/training/optimizations.md`.

sayakpaul commented on README.md:
```diff
@@ -128,540 +125,24 @@ eval $cmd
-> - Pass `--use_8bit_bnb` when launching training. Note that this is only applicable to Adam and AdamW optimizers.
-> - Do not perform validation/testing. This saves a significant amount of memory, which can be used to focus solely on training if you're on smaller VRAM GPUs.
-
-## Memory requirements
```
We don't need as have it here already:
https://github.com/a-r-r-o-w/finetrainers/tree/main/training

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+## Training
```
Would rename this file to cogvideox.md

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+## Training
+
```
Could you add the memory numbers to keep it consistent with other model docs? Will clean this up a bit more later but for now should be good

```
Training configuration: {
    "trainable parameters": 132120576,
    "total samples": 69,
    "train epochs": 1,
    "train steps": 10,
    "batches per device": 1,
    "total batches observed per epoch": 69,
    "train batch size": 1,
    "gradient accumulation steps": 1
}

| stage                         | memory_allocated  | max_memory_reserved |
|:-----------------------------:|:-----------------:|:-------------------:|
| after precomputing conditions |  8.880            | 8.941               |
| after precomputing latents    |  9.300            | 12.441              |
| before training start         | 10.622            | 20.701              |
| after epoch 1                 | 11.145            | 20.701              |
| before validation start       | 11.145            | 20.702              |
| after validation end          | 11.145            | 28.324              |
| after training end            | 11.144            | 11.592              |
```

a-r-r-o-w replied:
This is with the following settings btw:

```
LoRA with rank 128, batch size 1, gradient checkpointing, optimizer adamw, `49x480x720` resolutions, **with precomputation**:
```

sayakpaul replied:
Thanks! I just added it in 1f81197. Could you also let me know about the ckpt you used for this? Is it 1.0 5B or 1.5 5B?

sayakpaul replied:
I think this is 5b. So, updated accordingly.

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+from diffusers.utils import export_to_video
+
+pipe = DiffusionPipeline.from_pretrained(
+    "THUDM/CogVideoX-2b", torch_dtype=torch.float16
```
```suggestion
    "THUDM/CogVideoX-5b", torch_dtype=torch.bfloat16
```

I changed the defaults in the training script to CogVideoX-5b. This is because it is very hard to get anything good out of 2b and it has very low usage among finetuners IMO

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+pipe = DiffusionPipeline.from_pretrained(
+    "THUDM/CogVideoX-2b", torch_dtype=torch.float16
+).to("cuda")
++ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="ltxv-lora")
```
```suggestion
+ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="cogvideox-lora")
```

a-r-r-o-w commented on docs/training/cog.md:
````diff
@@ -0,0 +1,108 @@
+
+```diff
+import torch
+from diffusers import DiffusionPipeline
````
```suggestion
from diffusers import CogVideoXPipeline
```

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+from diffusers import DiffusionPipeline
+from diffusers.utils import export_to_video
+
+pipe = DiffusionPipeline.from_pretrained(
```
```suggestion
pipe = CogVideoXPipeline.from_pretrained(
```

a-r-r-o-w commented on docs/training/cog.md:
```diff
@@ -0,0 +1,108 @@
+    "THUDM/CogVideoX-2b", torch_dtype=torch.float16
+).to("cuda")
++ pipe.load_lora_weights("my-awesome-name/my-awesome-lora", adapter_name="ltxv-lora")
++ pipe.set_adapters(["ltxv-lora"], [0.75])
```
```suggestion
+ pipe.set_adapters(["cogvideox-lora"], [0.75])
```

a-r-r-o-w commented on docs/training/hunyuan.md:
```diff
@@ -0,0 +1,175 @@
+# Training
```
Would rename this file to `hunyuan_video.md` just to keep consistent with actual model names

a-r-r-o-w commented on docs/training/ltx.md:
```diff
@@ -0,0 +1,163 @@
+## Training
```
Would rename this file to `ltx_video.md` for same reason as above

a-r-r-o-w replied:
For all the model files, would maybe be good to add a `# {model_name}` header too

a-r-r-o-w commented on README.md:
```diff
@@ -120,548 +112,43 @@ cmd="accelerate launch --config_file accelerate_configs/uncompiled_2.yaml --gpu_
+## Notes
 
-Note: requires about `59` GB of VRAM when validation is performed.
+* The example training commands have the below thoughts in mind to prioritize memory-optimized runs:
+  * don't contain any arguments to enable validation inference.
+  * gradient checkpointing and gradient accumulation enabled.
+  * precompuation enabled.
```
I think okay to remove this part from main README

a-r-r-o-w commented on README.md:
````diff
@@ -120,548 +112,43 @@ cmd="accelerate launch --config_file accelerate_configs/uncompiled_2.yaml --gpu_
-).frames[0]
-export_to_video(output, "output.mp4", fps=15)
-```
+| Model Name | Tasks                | Ckpts Tested                                                                                                                                         | Min. GPU<br>VRAM | Comments                                                     |
````
Would remove the comments column here tbh. All models are supported with multi-resolution and multi-frames so far.

LTX is fast to train but it is extremely hard to teach it new styles tbh. I have had very limited success getting a good lora. Have not figured out the best settings yet but continuing to try.

sayakpaul replied:
Oh I kept the comments column for including anything we should note that couldn't be fit in the other columns. Anything that's out of the ordinary and is absolutely important for the users to know. 

WDYT?

a-r-r-o-w reviewed (commented):
Thanks for the incredible refactor!

sayakpaul commented:
@a-r-r-o-w thanks for the reviews! In the latest commits:

* I have addressed all your comments and left a question for you [here](https://github.com/a-r-r-o-w/finetrainers/pull/175#discussion_r1903263762). 
* Introduced an ultra-short section from the `docs/training/README.md` to reference the model-specific docs for easier navigation.

a-r-r-o-w reviewed (approved):
thanks!
